Unhandled exception during updating value for zookeeper-healthcheck-delay in Config Map

Step to reproduce:

1. Clone and open Strimzi sources https://github.com/strimzi/strimzi
2. Up cluster
`$ oc clucter up`
3. Login as admin
`$ oc login -u system:admin`
4. Deploy Cluster Controller
`$ oc create -f resources/openshift/cluster-controller-with-template.yaml`
5. Deploy Topic Controller
`$ oc create -f resources/openshift/topic-controller-with-template.yaml`
6. Login to OpenShift https://127.0.0.1:8443
7. Create "Apache Kafka (Ephemeral storage)"
7.1 Click "Add to project"
7.2 Click "Select from project"
7.3 Select "Apache Kafka (Ephemeral storage)" and click "Next"
7.4 Use default configurations:
- Number of Zookeper cluster nodes: 1
- Number of Kafka cluster nodes: 3
- Zookeeper healthcheck initial delay: 15
- Zookeeper healthcheck timeout: 5
7.5 Click "Create" and wait for the message "Apache Kafka (Ephemeral storage) has been created"
7.6 Open page "Pods" and wait when all pods will have status "Running"
8. Create Strimzi Topic Controller
8.1 Click "Add to project"
8.2 Click "Select from project"
8.3 Select "Strimzi Topic Controller" and click "Next"
8.4 Use default configurations
8.5 Click "Create" and wait for the message "Success Strimzi Topic Controller has been created."
8.6 Open page "Pods" and wait when all pods will have status "Running"
9. Open page "Config Maps" and open your cluster config
10. Edit value for zookeeper-healthcheck-delay
10.1 Click "Actions"
10.2 Click "Edit YAML"
10.3 Set new value (30) for variable zookeeper-healthcheck-delay and save
11. Open page "Pods" and open logs for strimzi-topic-controller

Actual result:
Log has an unhandled exception:
```
[2018-02-21 15:48:18,807] WARN <ClientCnxn :1162> [.local:2181)] Session 0x161b8fedf870003 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.NoRouteToHostException: No route to host
at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:361)
at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1141)
Feb 21, 2018 3:48:18 PM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.lang.RuntimeException: org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss for for path: /brokers/topics
at io.strimzi.controller.topic.TopicsWatcher.lambda$start$2(TopicsWatcher.java:67)
at io.strimzi.controller.topic.zk.ZkImpl.lambda$null$9(ZkImpl.java:347)
at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344)
at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss for for path: /brokers/topics
at org.apache.zookeeper.KeeperException.create(KeeperException.java:99)
at org.apache.zookeeper.KeeperException.create(KeeperException.java:51)
at io.strimzi.controller.topic.zk.ZkImpl.mapResult(ZkImpl.java:135)
at io.strimzi.controller.topic.zk.ZkImpl.lambda$children$10(ZkImpl.java:342)
at org.apache.zookeeper.ClientCnxn$EventThread.processEvent(ClientCnxn.java:601)
at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:505)
```

Additional details:
Docker version: 18.02.0-ce
OpenShift server version: v3.7.1

I tried to reproduce it and I see something different. Following the trace from the Topic Controller ...

```
[2018-03-27 07:58:01,749] INFO  <ClientCnxn  :1158> [.local:2181)] Unable to read additional data from server sessionid 0x162667351300003, likely server has closed socket, closing socket connection and attempting reconnect
--
  | [2018-03-27 07:58:03,489] INFO  <ClientCnxn  :1032> [.local:2181)] Opening socket connection to server my-cluster-zookeeper.myproject.svc.cluster.local/172.30.85.68:2181. Will not attempt to authenticate using SASL (unknown error)
  | [2018-03-27 07:58:21,867] WARN  <ClientCnxn  :1108> [.local:2181)] Client session timed out, have not heard from server in 20018ms for sessionid 0x162667351300003
  | [2018-03-27 07:58:21,868] INFO  <ClientCnxn  :1156> [.local:2181)] Client session timed out, have not heard from server in 20018ms for sessionid 0x162667351300003, closing socket connection and attempting reconnect
  | Mar 27, 2018 7:58:21 AM io.vertx.core.impl.ContextImpl
  | SEVERE: Unhandled exception
  | java.lang.RuntimeException: org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss for for path: /brokers/topics
  | at io.strimzi.controller.topic.TopicsWatcher.lambda$start$2(TopicsWatcher.java:67)
  | at io.strimzi.controller.topic.zk.ZkImpl.lambda$null$9(ZkImpl.java:348)
  | at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344)
  | at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
  | at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
  | at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
  | at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
  | at java.lang.Thread.run(Thread.java:748)
  | Caused by: org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss for for path: /brokers/topics
  | at org.apache.zookeeper.KeeperException.create(KeeperException.java:99)
  | at org.apache.zookeeper.KeeperException.create(KeeperException.java:51)
  | at io.strimzi.controller.topic.zk.ZkImpl.mapResult(ZkImpl.java:135)
  | at io.strimzi.controller.topic.zk.ZkImpl.lambda$children$10(ZkImpl.java:343)
  | at org.apache.zookeeper.ClientCnxn$EventThread.processEvent(ClientCnxn.java:601)
  | at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:505)
  |  
  | [2018-03-27 07:58:23,126] INFO  <ClientCnxn  :1032> [.local:2181)] Opening socket connection to server my-cluster-zookeeper.myproject.svc.cluster.local/172.30.85.68:2181. Will not attempt to authenticate using SASL (unknown error)
  | [2018-03-27 07:58:43,146] WARN  <ClientCnxn  :1108> [.local:2181)] Client session timed out, have not heard from server in 21177ms for sessionid 0x162667351300003
  | [2018-03-27 07:58:43,146] INFO  <ClientCnxn  :1156> [.local:2181)] Client session timed out, have not heard from server in 21177ms for sessionid 0x162667351300003, closing socket connection and attempting reconnect
  | [2018-03-27 07:58:44,573] INFO  <ClientCnxn  :1032> [.local:2181)] Opening socket connection to server my-cluster-zookeeper.myproject.svc.cluster.local/172.30.85.68:2181. Will not attempt to authenticate using SASL (unknown error)
  | [2018-03-27 07:58:44,574] INFO  <ClientCnxn  :876> [.local:2181)] Socket connection established to my-cluster-zookeeper.myproject.svc.cluster.local/172.30.85.68:2181, initiating session
  | [2018-03-27 07:58:44,580] INFO  <ClientCnxn  :1158> [.local:2181)] Unable to read additional data from server sessionid 0x162667351300003, likely server has closed socket, closing socket connection and attempting reconnect
```

In any case the worst thing is that when the ZK pod comes up, the Kafka brokers aren't able to connect to it and the cluster is unusable.

The ZK log has a lot of these lines as attempts from Kafka brokers to connect ...

```
2018-03-27 08:02:44,504 INFO Accepted socket connection from /172.17.0.7:38840 (org.apache.zookeeper.server.NIOServerCnxnFactory) [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181]
--
  | 2018-03-27 08:02:44,505 DEBUG Session establishment request from client /172.17.0.7:38840 client's lastZxid is 0x47 (org.apache.zookeeper.server.ZooKeeperServer) [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181]
  | 2018-03-27 08:02:44,505 INFO Refusing session request for client /172.17.0.7:38840 as it has seen zxid 0x47 our last zxid is 0x2 client must try another server (org.apache.zookeeper.server.ZooKeeperServer) [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181]
```

each Kafka broker has log lines as the following trying to reconnect ...

```
2018-03-27 08:03:38,141 INFO Opening socket connection to server my-cluster-zookeeper.myproject.svc.cluster.local/172.30.85.68:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(my-cluster-zookeeper.myproject.svc.cluster.local:2181)]
--
  | 2018-03-27 08:03:38,141 INFO Socket connection established to my-cluster-zookeeper.myproject.svc.cluster.local/172.30.85.68:2181, initiating session (org.apache.zookeeper.ClientCnxn) [main-SendThread(my-cluster-zookeeper.myproject.svc.cluster.local:2181)]
  | 2018-03-27 08:03:38,141 INFO Unable to read additional data from server sessionid 0x162667351300000, likely server has closed socket, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn) [main-SendThread(my-cluster-zookeeper.myproject.svc.cluster.local:2181)]
```

Of course trying to log into the ZK using the shell tool, the `ls /` command reports no znodes related to the Kafka cluster (because brokers weren't able to connect and no znodes were created of course).

--- a/documentation/adoc/cluster-controller.adoc
+++ b/documentation/adoc/cluster-controller.adoc
@@ -135,9 +135,6 @@ data:
     }
 ----
 
-WARNING: If the zookeeper cluster is deployed using "ephemeral" storage, the Kafka brokers can have problems dealing with
-Zookeeper node restarts which could happen via updates in the cluster ConfigMap.
-
 The resources created by the cluster controller into the Kubernetes/OpenShift cluster will be the following :
 
 * `[cluster-name]-zookeeper` StatefulSet which is in charge to create the Zookeeper node pods
@@ -172,6 +169,9 @@ The "ephemeral" storage is really simple to configure and the related JSON strin
 { "type": "ephemeral" }
 
 ----
+
+WARNING: If the Zookeeper cluster is deployed using "ephemeral" storage, the Kafka brokers can have problems dealing with
+Zookeeper node restarts which could happen via updates in the cluster ConfigMap.
 
 In case of "persistent-claim" type the following fields can be provided as well :
 

--- a/documentation/adoc/cluster-controller.adoc
+++ b/documentation/adoc/cluster-controller.adoc
@@ -135,8 +135,8 @@ data:
     }
 ----
 
-WARNING: If the cluster is deployed using the "ephemeral" storage, the Kafka brokers could have problems to deal with
-Zookeeper nodes restarts which could happen on configuration updates in the cluster ConfigMap.
+WARNING: If the zookeeper cluster is deployed using "ephemeral" storage, the Kafka brokers can have problems dealing with
+Zookeeper node restarts which could happen via updates in the cluster ConfigMap.
 
 The resources created by the cluster controller into the Kubernetes/OpenShift cluster will be the following :
 

--- a/documentation/adoc/cluster-controller.adoc
+++ b/documentation/adoc/cluster-controller.adoc
@@ -135,6 +135,9 @@ data:
     }
 ----
 
+WARNING: If the cluster is deployed using the "ephemeral" storage, the Kafka brokers could have problems to deal with
+Zookeeper nodes restarts which could happen on configuration updates in the cluster ConfigMap.
+
 The resources created by the cluster controller into the Kubernetes/OpenShift cluster will be the following :
 
 * `[cluster-name]-zookeeper` StatefulSet which is in charge to create the Zookeeper node pods

--- a/topic-controller/src/main/java/io/strimzi/controller/topic/TopicsWatcher.java
+++ b/topic-controller/src/main/java/io/strimzi/controller/topic/TopicsWatcher.java
@@ -64,7 +64,8 @@ class TopicsWatcher {
                 return;
             }
             if (childResult.failed()) {
-                throw new RuntimeException(childResult.cause());
+                LOGGER.error("Error on zone {} children", TOPICS_ZNODE, childResult.cause());
+                return;
             }
             List<String> result = childResult.result();
             LOGGER.debug("znode {} now has children {}, previous children {}", TOPICS_ZNODE, result, this.children);
@@ -106,7 +107,8 @@ class TopicsWatcher {
 
         }).children(TOPICS_ZNODE, childResult -> {
             if (childResult.failed()) {
-                throw new RuntimeException(childResult.cause());
+                LOGGER.error("Error on zone {} children", TOPICS_ZNODE, childResult.cause());
+                return;
             }
             List<String> result = childResult.result();
             LOGGER.debug("Setting initial children {}", result);

--- a/topic-controller/src/main/java/io/strimzi/controller/topic/TopicsWatcher.java
+++ b/topic-controller/src/main/java/io/strimzi/controller/topic/TopicsWatcher.java
@@ -64,7 +64,7 @@ class TopicsWatcher {
                 return;
             }
             if (childResult.failed()) {
-                LOGGER.error("Error on zone {} children", TOPICS_ZNODE, childResult.cause());
+                LOGGER.error("Error on znode {} children", TOPICS_ZNODE, childResult.cause());
                 return;
             }
             List<String> result = childResult.result();
@@ -107,7 +107,7 @@ class TopicsWatcher {
 
         }).children(TOPICS_ZNODE, childResult -> {
             if (childResult.failed()) {
-                LOGGER.error("Error on zone {} children", TOPICS_ZNODE, childResult.cause());
+                LOGGER.error("Error on znode {} children", TOPICS_ZNODE, childResult.cause());
                 return;
             }
             List<String> result = childResult.result();

